cpp/src/solver/lars_impl.cuh

```diff
@@ -0,0 +1,717 @@
+    if (G == nullptr) {
+      // Calculate a = X.T[:,n_active:] * u                              (2.11)
+      math_t one = 1;
+      math_t zero = 0;
+      CUBLAS_CHECK(raft::linalg::cublasgemv(
+        handle.get_cublas_handle(), CUBLAS_OP_T, n_rows, n_inactive, &one,
+        X + n_active * ld_X, ld_X, u, 1, &zero, a_vec, 1, stream));
+    } else {
+      // Calculate a = X.T[:,n_A:] * u = X.T[:, n_A:] * X[:,:n_A] * ws
+      //             = G[n_A:,:n_A] * ws                                 (2.11)
+      math_t one = 1;
+      math_t zero = 0;
+      CUBLAS_CHECK(raft::linalg::cublasgemv(
+        handle.get_cublas_handle(), CUBLAS_OP_N, n_inactive, n_active, &one,
+        G + n_active, ld_G, ws, 1, &zero, a_vec, 1, stream));
+    }
+    const math_t tiny = std::numeric_limits<math_t>::min();
+    const math_t huge = std::numeric_limits<math_t>::max();
+    //
+    // gamma = min^+_{j \in inactive} {(Cmax - cor_j) / (A-a_j),
+    //                                 (Cmax + cor_j) / (A+a_j)}         (2.13)
+    auto map = [Cmax, A, tiny, huge] __device__(math_t c, math_t a) -> math_t {
+      math_t tmp1 = (Cmax - c) / (*A - a + tiny);
+      math_t tmp2 = (Cmax + c) / (*A + a + tiny);
+      // We consider only positive elements while we search for the minimum
+      math_t val = (tmp1 > 0) ? tmp1 : huge;
+      if (tmp2 > 0 && tmp2 < val) val = tmp2;
+      return val;
+    };
+    raft::linalg::mapThenReduce(gamma, n_inactive, huge, map, cub::Min(),
+                                stream, cor + n_active, a_vec);
+  }
+}
+
+/**
+ * @brief Initialize for Lars training.
+ */
+template <typename math_t, typename idx_t>
+void larsInit(const raft::handle_t& handle, const math_t* X, idx_t n_rows,
+              idx_t n_cols, idx_t ld_X, const math_t* y, math_t* Gram,
+              idx_t ld_G, MLCommon::device_buffer<math_t>& U_buffer, math_t** U,
+              idx_t* ld_U, MLCommon::host_buffer<idx_t>& indices,
+              MLCommon::device_buffer<math_t>& cor,
+              MLCommon::device_buffer<math_t>& sign, int* max_iter,
+              math_t* coef_path, cudaStream_t stream) {
+  if (n_cols < *max_iter) {
+    *max_iter = n_cols;
+  }
+  if (Gram == nullptr) {
+    const idx_t align_bytes = 256;
+    *ld_U = raft::alignTo<idx_t>(*max_iter, align_bytes);
+    try {
+      U_buffer.resize((*ld_U) * (*max_iter), stream);
+    } catch (std::bad_alloc) {
+      THROW(
+        "Not enough GPU memory! The memory usage depends quadraticaly on the "
+        "n_nonzero_coefs parameter, try to decrease it!");
+    }
+    *U = U_buffer.data();
+  } else {
+    // Set U as G. During the solution in larsFit, the Cholesky factorization
+    // U will overwrite G.
+    *U = Gram;
+    *ld_U = ld_G;
+  }
+  std::iota(indices.data(), indices.data() + n_cols, 0);
+
+  math_t one = 1;
+  math_t zero = 0;
+  // Set initial correlation to X.T * y
+  // TODO: this returns NaNs if n_rows >= 65536 and math_t = fp32
```

Comment:
Do you have a repro for this? If yes, can you file a bug against cublas for this one?

Reply:
Further investigations showed that it is not a cuBLAS error, but a tricky problem with the input data. I have opened issue #3189 to track this. Additionally, I will disable the fp32 solver path until the issue is resolved.